README
Welcome to CW Trainer!

Legal Stuff:

This application and its source code are hereby released under the
GPL, Version 2. Please find the enclosed LICENSE for more information
on the GPL.

Initial release date is 2/15/2002 by John Wagner, N1QO -
john@wagner-usa.net

The program uses some Source Code from Ray Goff, G4FON - notably the
EncodeSineWave function. That code is used and released under the GPL
by permission from Ray. It should also be noted that Ray gave me the 
source code to his windows program and from that I was inspired to 
write this program.

The files lcdrange.h and lcdrange.cpp are from the QT Tutorial.
These files may be removed at a later date.

********** BEFORE YOU SUBMIT REQUESTS FOR FEATURES ********** 

Please read the file TODO - the early versions of this program are
more or less working prototypes intended to get something [anything]
out there for Linux to help people learn CW. Numerous features are
planned as well as a general unfudging of the source code. Again, the
early versions are out here just so people have something to use.

Purpose:

This application is intended to teach Morse Code (CW). 

It supports the Koch method;

	Teach at high speed, starting with two characters. When
	those character is recognized 90% of the time then add
	another character. Repeat until using all characters and 90%
	correctness is achieved... then speed up and start over!

Koch Method Resources:
Dave Finley, N1IRZ - http://www.sdc.org/~finley/
Windows software by Ray Goff, G4FON - http://www.qsl.net/g4fon
Go to http://www.google.com/ - type in "koch method" - tons of pages

It supports the Farnsworth method;

	Fast character speed (typically 20wpm) but wide character
	space (start at 5 wpm). Can use all characters at the same
	time or use with the Koch method above.

Farnsworth Method Resources:
Go to http://www.google.com/ - type in "farnsworth cw" - tons of pages

Requires QT Version >= 3.x multi-thread library (Currently developed against 
libqt3-mt-dev v3.3.4-3) and a Linux system with /dev/dsp sound support. 

Building and running:

Unpack the archive: 

tar zxvf cwtrainer.tar.gz

Ensure you have your QTDIR environment variable set;

export QTDIR=/usr/share/qt3

(yours may be different, install the *qt3*-dev package for your system and look
for a directory with this structure:
qt<version>/
           bin/
	   doc/
	   include/
	   lib/
	   mkspecs/
	   plugins/
	   )

type: make

You should now have an executable called "cwtrainer"

Run it!: ./cwtrainer

There is no installation target yet. In the short term you may want to put
cwtrainer into your path if you don't want to give the path to this folder or
change directories each time you run it. Your $HOME/bin directory would be 
great for this, or /usr/local/bin if you want a site-wide location. 

A cwtrainer.pro file is included for tmake users who want to hack on
the program using qmake.

The "Load File..." will read a text file and play it as CW. At the
moment you will note that once you hit the "Stop CW" or "Start CW"
button, you will need to reload the file if you want to play it again.
A sample QSO is included in the file "qso.txt". Also note that Prosigns
are supported in the text file, append a ^ before the prosign;

Examples: ^AR ^BK ^BT ^SK

Maintainer:
John Wagner, N1QO - john@wagner-usa.net (original author)
Jacob Anawalt, KD7YKO - jlanawalt@users.sourceforge.net

Thanks to:
Ray Goff, G4FON - Source code to Koch
Karl Larsen, K5DI - testing and suggestions

History:
See CHANGES

Similar Projects:
It looks like a few other projects for learning CW have been posted to sf.net
since Feb 2002. If cwtrainer isn't at all what you are looking for, you may
want to check them out.

cw-trainer
* http://freshmeat.net/projects/cwtrainer/
* http://sourceforge.net/projects/cw-trainer/

cwtutor
* http://sourceforge.net/projects/cwtutor/
** Summary is almost exactly the same as this package's summary, but it is
   written quite differently.

cwtext
* http://freshmeat.net/projects/cwtext/
* http://cwtext.sourceforge.net/

cwstation
* http://sourceforge.net/projects/cwstation



CHANGES
cwtrainer (0.0.1)
  * First public release
 -- John Wagner <john@wagner-usa.net>  Wed,  13 Feb 2002

cwtrainer (0.0.2)
  * Private release for those using a newer compiler
 -- John Wagner <john@wagner-usa.net>

cwtrainer (0.0.3)
  * Second public release.
  * Code compatibility changes for gcc 3.3.
 -- Jacob Anawalt <jlanawalt@users.sourceforge.net>

cwtrainer (0.1.0)
  * Targeting QT3:
  ** Switched from tmake to qmake
  ** Building against qt-mt instead of qt
  * Updated About box.
 -- Jacob Anawalt <jlanawalt@users.sourceforge.net>


cwtest.h
/************************************************************************
 *  THIS CODE IS RELEASED UNDER THE
 *  GNU GENERAL PUBLIC LICENSE
 *  Version 2, June 1991
 *
 *  See LICENSE for more information
 *
 ************************************************************************/

#ifndef __CW__TEST__H__
#define __CW__TEST__H__

#include <map>
#include <string>

class CWTest
{
 private:
  struct CWTestChar
  {    
    std::string m_morse;
    bool m_inuse;
    CWTestChar(const char* morse, bool inuse) :
      m_morse(morse), m_inuse(inuse){}
    CWTestChar(){}
  };

  typedef std::map<char, CWTest::CWTestChar> CWCharMap;
  CWCharMap m_cw;

  CWTest()
  {
    m_cw['A'] = CWTestChar(".-", true);
    m_cw['B'] = CWTestChar("-...", true);
    m_cw['C'] = CWTestChar("-.-.", true);
    m_cw['D'] = CWTestChar("-..", true);
    m_cw['E'] = CWTestChar(".", true);
    m_cw['F'] = CWTestChar("..-.", true);
    m_cw['G'] = CWTestChar("--.", true);
    m_cw['H'] = CWTestChar("....", true);
    m_cw['I'] = CWTestChar("..", true);
    m_cw['J'] = CWTestChar(".---", true);
    m_cw['K'] = CWTestChar("-.-", true);
    m_cw['L'] = CWTestChar(".-..", true);
    m_cw['M'] = CWTestChar("--", true);
    m_cw['N'] = CWTestChar("-.", true);
    m_cw['O'] = CWTestChar("---", true);
    m_cw['P'] = CWTestChar(".--.", true);
    m_cw['Q'] = CWTestChar("--.-", true);
    m_cw['R'] = CWTestChar(".-.", true);
    m_cw['S'] = CWTestChar("...", true);
    m_cw['T'] = CWTestChar("-", true);
    m_cw['U'] = CWTestChar("..-", true);
    m_cw['V'] = CWTestChar("...-", true);
    m_cw['W'] = CWTestChar(".--", true);
    m_cw['X'] = CWTestChar("-..-", true);
    m_cw['Y'] = CWTestChar("-.--", true);
    m_cw['Z'] = CWTestChar("--..", true);
    m_cw['1'] = CWTestChar(".----", true);
    m_cw['2'] = CWTestChar("..---", true);
    m_cw['3'] = CWTestChar("...--", true);
    m_cw['4'] = CWTestChar("....-", true);
    m_cw['5'] = CWTestChar(".....", true);
    m_cw['6'] = CWTestChar("-....", true);
    m_cw['7'] = CWTestChar("--...", true);
    m_cw['8'] = CWTestChar("---..", true);
    m_cw['9'] = CWTestChar("----.", true);
    m_cw['0'] = CWTestChar("-----", true);
    m_cw['.'] = CWTestChar(".-.-.-", true);
    m_cw[','] = CWTestChar("--..--", true);
    m_cw['?'] = CWTestChar("..--..", true);
    m_cw['/'] = CWTestChar("-..-.", true);
    m_cw[' '] = CWTestChar(" ", true);
  }

  static CWTest& GetChars()
  {
    static CWTest cw;
    return cw;
  }

 public:
  static bool CharInUse(char ch)
    {
      CWCharMap::const_iterator iter = GetChars().m_cw.find(ch);

      if(iter == GetChars().m_cw.end())
	{
	  return false;
	}

      return (*iter).second.m_inuse;
    }

  static void SetCharInUse(char ch, bool state)
    {
      CWCharMap::const_iterator iter = GetChars().m_cw.find(ch);

      if(iter == GetChars().m_cw.end())
	{
	  return;
	}

      GetChars().m_cw[ch].m_inuse = state;
    }
  
  static const std::string& GetCW(char ch)
    {
      CWCharMap::const_iterator iter = GetChars().m_cw.find(ch);

      if(iter == GetChars().m_cw.end())
	{
	  return GetChars().m_cw[' '].m_morse;
	}

      return GetChars().m_cw[ch].m_morse;
    }
};


#endif


layout.h
/************************************************************************
 *  THIS CODE IS RELEASED UNDER THE
 *  GNU GENERAL PUBLIC LICENSE
 *  Version 2, June 1991
 *
 *  See LICENSE for more information
 *
 ************************************************************************/

#ifndef __MY__LAYOUT__H__
#define __MY__LAYOUT__H__

#include <qwidget.h>
#include <qtimer.h>
//#include <qbuttongroup.h>
#include <qtextedit.h>
#include "lcdrange.h"
//#include <QLCDNumber>
#include "cwsound.h"
#include <QGroupBox>
#include <QGridLayout>

class MyWidget : public QWidget
{
  Q_OBJECT

private:
  CWSoundMachine* m_cw;
  QTimer *m_timer;
  QGroupBox *m_lettergroup;
  QGridLayout *m_letterlayout;
  QTextEdit* m_cwtext;
  LCDRange* m_wpm;
  LCDRange* m_spacing;

  QString m_charpipe;

  //chars in suggest Koch order
  static const char * m_cwchars;
  char m_charsInTest[41];
  int m_numInTest;

  char m_lastchar;
  int m_numingroup;
  int m_charspace;
  int m_wordspace;
public:
  MyWidget( QWidget *parent=0 );
  ~MyWidget()
    {
      delete m_cw;
      delete m_timer;
      delete m_lettergroup;
      delete m_cwtext;
    }

protected slots:
    void StartCW();
    void EndCW();
    void About();
    void ClearAll();
    void SelectAll();
    void NewSpacing(int);
    void NewSpeed(int);
    void ReadFile();

    void playNextChar();
    // utility functions
    void ConvertToPosition( int i, int *row, int *column);
    int ConvertFromPosition( int row, int column );
};

#endif


main.cpp
/************************************************************************
 *  THIS CODE IS RELEASED UNDER THE
 *  GNU GENERAL PUBLIC LICENSE
 *  Version 2, June 1991
 *
 *  See LICENSE for more information
 *
 ************************************************************************/

/****************************************************************
 **
 ** mtrainer
 ** cw training software
 **
 ** Based on and includes some of Ray Goff, G4FON's "Koch" for windows
 ** 
 **
 ** 
 **
 ****************************************************************/

#include <qapplication.h>
#include "layout.h"

int main( int argc, char **argv )
{
  QApplication a( argc, argv );

  //MyWidget w(NULL, "CW Trainer");
  MyWidget w(NULL) ;
//  a.setMainWidget( &w );
  w.setMinimumSize(640, 480);
  w.show();
  return a.exec();
}


cwsound.h
/************************************************************************
 *  THIS CODE IS RELEASED UNDER THE
 *  GNU GENERAL PUBLIC LICENSE
 *  Version 2, June 1991
 *
 *  See LICENSE for more information
 *
 ************************************************************************/

#ifndef __CW__SOUND__H__
#define __CW__SOUND__H__

void makesound();

class CWSoundMachine
{
 private:
  int m_dsp; //handle to dsp device
  int m_speed;
  int m_freq;
  unsigned char *m_ditbuffer;
  unsigned char *m_dahbuffer;
  int m_ditlen;
  int m_dahlen;
  int m_charspacelen;
  int m_wordspace;

  void BuildBuffers();

  //prevent copies
  CWSoundMachine(const CWSoundMachine& cw){};
 public:
  CWSoundMachine(int speed, int charspacelen, int freq);
  ~CWSoundMachine();

  void AdjustSpeed(int speed);

  void PlayCWNote(const char *note);
  int CharSpaceLen(){return m_charspacelen;}
  int WordSpace(){return m_wordspace;}
};

#endif


lcdrange.h


/****************************************************************
**
** Definition of LCDRange class, Qt tutorial 7
**
****************************************************************/

/*#ifndef LCDRANGE_H
#define LCDRANGE_H

//#include <qvbox.h>
#include<QVBoxLayout>

class QSlider;

//class LCDRange : public QVBox
class LCDRange : public QVBoxLayout
{
    Q_OBJECT
public:
    LCDRange(int min, int max, QWidget *parent=0, const char *name=0 );

    int value() const;

public slots:
    void setValue( int );

signals:
    void valueChanged( int );

private:
    QSlider *slider;
};

#endif // LCDRANGE_H*/

// http://www.greyc.ensicaen.fr/ensicaen/Docs/Qt4/tutorial-t14-lcdrange-h.html

/****************************************************************************
 **
 ** Copyright (C) 2005-2006 Trolltech ASA. All rights reserved.
 **
 ** This file is part of the example classes of the Qt Toolkit.
 **
 ** This file may be used under the terms of the GNU General Public
 ** License version 2.0 as published by the Free Software Foundation
 ** and appearing in the file LICENSE.GPL included in the packaging of
 ** this file.  Please review the following information to ensure GNU
 ** General Public Licensing requirements will be met:
 ** http://www.trolltech.com/products/qt/opensource.html
 **
 ** If you are unsure which license is appropriate for your use, please
 ** review the following information:
 ** http://www.trolltech.com/products/qt/licensing.html or contact the
 ** sales department at sales@trolltech.com.
 **
 ** This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
 ** WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 **
 ****************************************************************************/

 #ifndef LCDRANGE_H
 #define LCDRANGE_H

 #include <QWidget>

 class QLabel;
 class QSlider;

 class LCDRange : public QWidget
 {
     Q_OBJECT

 public:
     LCDRange(QWidget *parent = 0);
     LCDRange(const QString &text, QWidget *parent = 0);

     int value() const;
     QString text() const;

 public slots:
     void setValue(int value);
     void setRange(int minValue, int maxValue);
     void setText(const QString &text);

 signals:
     void valueChanged(int newValue);

 private:
     void init();

     QSlider *slider;
     QLabel *label;
 };

 #endif


cwsound.cpp
/************************************************************************
 *  THIS CODE IS RELEASED UNDER THE
 *  GNU GENERAL PUBLIC LICENSE
 *  Version 2, June 1991
 *
 *  See LICENSE for more information
 *
 ************************************************************************/

#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <stdlib.h>
#include <stdio.h>
#include <linux/soundcard.h>
#include <ctype.h>
#include <math.h>
#include <string.h>

#include "cwsound.h"


void play_note(unsigned char* note, int fd_dsp, int dur)
{
  int status;
  status = write(fd_dsp, note, dur); /* play it back */
  status = ioctl(fd_dsp, SOUND_PCM_SYNC, 0); 
}

#define SAMPLE_RATE		11025

void EncodeSineWave ( unsigned char *Into, int Len, int Pitch )
{
  int x;
  double fAngle;
  unsigned char a;
  int Amplitude;

  fAngle = 0;

  //	this routine fills the region of the tone buffer with a sine wave...
  for ( Amplitude = x = 0; x < Len; x++ )
    {
      //	factor 1 & 2 set the amplitude of the sine wave... By
      //	slowly building it up we avoid any nasty key clicks...

      a = (unsigned char)( Amplitude * sin ( fAngle ));
      a += (unsigned char)128;

      //	give the waveform a gentle start and finish....

      if ( x < 64 )
	{
	  if ( Amplitude < 100 )
	    Amplitude += 2;
	}
      else if ( x >= ( Len - 64 ))
	{
	  if ( Amplitude > 0 )
	    Amplitude -= 2;
	}

      //	poke the byte into the buffer...

      *Into++ = a;

      //	advance the phase of the sine wave...

      fAngle += ( 2 * M_PI * Pitch ) / SAMPLE_RATE;

      //	keep it in the range 0..2*PI

      if ( fAngle > ( 2 * M_PI ))
	fAngle -= ( 2 * M_PI );
    }
}


CWSoundMachine::CWSoundMachine(int speed, int charspacelen, int freq) :
  m_speed(speed),
  m_freq(freq),
  m_charspacelen(charspacelen)
{
  int maxlen = ((SAMPLE_RATE * 1200) / 1000) * 3 ;
  m_ditbuffer = new unsigned char[maxlen];
  m_dahbuffer = new unsigned char[maxlen];


  /* open sound device */
  m_dsp = open("/dev/dsp", O_WRONLY, 0);
  if (m_dsp < 0) {
    perror("open of /dev/dsp failed");
    _exit(1);
  }
  int value = AFMT_U8;
  ioctl (m_dsp, SNDCTL_DSP_SETFMT, &value);

  int n_fragments = 16; /* number of fragments */
  int fragment_size = 8; /* a buffersize of 2^8 = 256 bytes */

  value = (n_fragments << 16) | fragment_size;
  ioctl (m_dsp, SNDCTL_DSP_SETFRAGMENT, &value);

  value = SAMPLE_RATE; /* frequency */
  ioctl (m_dsp, SNDCTL_DSP_STEREO, &value);

  value = 8; /* number of bits */
  ioctl (m_dsp, SNDCTL_DSP_SAMPLESIZE, &value);

  value = 0; /* mono = 0, stereo = 1 */
  ioctl (m_dsp, SNDCTL_DSP_STEREO, &value);

  BuildBuffers();
}


CWSoundMachine::~CWSoundMachine()
{
  delete [] m_ditbuffer;
  delete [] m_dahbuffer;

  close(m_dsp);
}

void CWSoundMachine::AdjustSpeed(int speed)
{
  m_speed = speed;
  BuildBuffers();
}

void CWSoundMachine::BuildBuffers()
{
//   if(m_ditbuffer)
//     delete [] m_ditbuffer;
//   if(m_dahbuffer)
//     delete [] m_dahbuffer;

  m_ditlen = (( SAMPLE_RATE * ( 1200 / m_speed )) / 1000 );
  m_ditlen /= 2;
  m_dahlen = m_ditlen*3;
  m_wordspace = m_dahlen;
  m_charspacelen = m_ditlen;

  //printf("ws: %d - cs: %d\n", m_wordspace, m_charspacelen);

  EncodeSineWave(m_ditbuffer, m_ditlen, m_freq);
  EncodeSineWave(m_dahbuffer, m_dahlen, m_freq);

  memset(m_ditbuffer + m_ditlen, 128, m_ditlen);
  memset(m_dahbuffer + m_dahlen, 128, m_ditlen);

}

void CWSoundMachine::PlayCWNote(const char *note)
{
  for(int i=0 ; note[i] ; i++)
    {
      int dur = note[i] == '.' ? m_ditlen : m_dahlen;
      unsigned char *buf = (note[i] == '.' ? m_ditbuffer : m_dahbuffer);
      play_note(buf, m_dsp, dur+m_ditlen);
    }
}


layout.cpp
/************************************************************************
 *  THIS CODE IS RELEASED UNDER THE
 *  GNU GENERAL PUBLIC LICENSE
 *  Version 2, June 1991
 *
 *  See LICENSE for more information
 *
 ************************************************************************/

#include <time.h>

#include <qpushbutton.h>
#include <qlabel.h>
//#include <qgrid.h>
#include <QGridLayout>
#include <qlayout.h>
#include <qmessagebox.h>
#include <qcheckbox.h>
#include <qfont.h>
#include <qfiledialog.h>
#include <qfile.h>
#include <qtextstream.h>
#include "layout.h"
#include "cwsound.h"

#include "cwtest.h"
#include "lcdrange.h"
#include <QTextEdit>
#include <QGroupBox>
#include <iostream>

#define CODESPEED_MIN 15
#define CODESPEED_MAX 40
#define CODESPEED_DEFAULT 20

#define SPACING_MIN 5
#define SPACING_MAX 40
#define SPACING_DEFAULT 20

#define CHARACTER_GRID_ROWS 8

const char * MyWidget::m_cwchars = "KMRSUAPTLOWI.NJEFOY,VG5/Q9ZH38B?427C1D6X";

MyWidget::MyWidget( QWidget *parent )
  : QWidget( parent )
{
  srand(time(NULL));
  m_cw = new CWSoundMachine(20, 1, 600);
  m_lastchar = '&';

  //QGridLayout *grid = new QGridLayout(this, 2, 2, 10);
  QGridLayout *grid = new QGridLayout(this);
  //grid->setSpacing(10);

  //m_wpm = new LCDRange(CODESPEED_MIN, 40, this);
  LCDRange *m_wpm = new LCDRange(this);
  m_wpm->setRange(CODESPEED_MIN, CODESPEED_MAX);
  //m_wpm->setValue(20);
  m_wpm->setValue(CODESPEED_DEFAULT);

  connect(m_wpm, SIGNAL(valueChanged(int)),
	  SLOT(NewSpeed(int)) );


  //QLabel *label = new QLabel ( m_wpm, "CW Character Speed", this);
  QLabel *label = new QLabel ( "CW Character Speed", this);
  //label->setBuddy(m_wpm);
  grid->addWidget(label, 0, 0);
  grid->addWidget(m_wpm, 1, 0, Qt::Alignment(Qt::AlignLeft) );

  //m_spacing = new LCDRange(5, 40, this);
  m_spacing = new LCDRange(this);
  m_spacing->setRange(SPACING_MIN, SPACING_MAX);
  //m_spacing->setValue(20);
  m_spacing->setValue(SPACING_DEFAULT);

  connect(m_spacing, SIGNAL(valueChanged(int)),
	  SLOT(NewSpacing(int)) );

  grid->addWidget(m_spacing, 1, 1);
  //label = new QLabel (m_wpm, "Character Spacing", this);
  label = new QLabel ("Character Spacing");
  grid->addWidget(label, 0, 1);

  //m_cwtext = new QMultiLineEdit(this);
  m_cwtext = new QTextEdit(this);
  m_cwtext->setFont(QFont( "Courier", 18, QFont::Bold ));

  m_cwtext->setReadOnly(true);
  //m_cwtext->setWordWrap(QMultiLineEdit::WidgetWidth);
  m_cwtext->setWordWrapMode(QTextOption::WordWrap);
  //grid->addMultiCellWidget(m_cwtext, 2, 2, 0, 1);
  grid->addWidget(m_cwtext, 2, 0, 1, 2);

  QPushButton *start = new QPushButton( "Start CW", this);
  QPushButton *stop = new QPushButton( "Stop CW", this);

  connect( start, SIGNAL(clicked()),
	   SLOT(StartCW()) );

  connect( stop, SIGNAL(clicked()),
	   SLOT(EndCW()) );

  grid->addWidget(start, 3, 0);
  grid->addWidget(stop, 3, 1);

  const char *ptr = m_cwchars;
  //m_lettergroup = new QGroupBox(5, Horizontal, "Chars", this) ;
  m_lettergroup = new QGroupBox("Chars", this) ;
  m_lettergroup->setAlignment( Qt::Alignment(Qt::AlignHCenter) );
  m_letterlayout = new QGridLayout;
  //int chars_in_row = QString(ptr).length() / CHARACTER_GRID_ROWS;
  for(int i = 0 ; ptr[i] ; i++)
    {
      //QCheckBox *cb = new QCheckBox(QString(QChar(ptr[i])), m_lettergroup);
      QCheckBox *cb = new QCheckBox(QString(QChar(ptr[i])), m_lettergroup);
      //m_lettergroup->addButton( cb );

      cb->setChecked(CWTest::CharInUse(ptr[i]));
     // guessing at coordinates.
     // QButtonGroup cannot be added to QGridLayout
     // http://www.qtcentre.org/threads/42754-Can-t-add-QButtonGroup-to-QGridLayout
     //grid->addWidget(cb , 4, i, Qt::Alignment(Qt::AlignLeft));
     //int row = 0;
     //if( i != 0) // take care of divide-by-zero
      //  row = i / chars_in_row;
     //int column  = i - row*chars_in_row;
     int row, column;
     ConvertToPosition( i, &row, &column); // modifies row and column
     m_letterlayout->addWidget(cb, row, column);

//     m_lettergroup->insert(cb);
    }
  m_lettergroup->setLayout(m_letterlayout);

  //grid->addWidget(m_lettergroup, 4, 0);
    //grid->addWidget(m_letterlayout, 4, 0);
  //grid->addLayout(m_letterlayout, 4, 0);
  grid->addWidget(m_lettergroup, 4, 0);

  //QButtonGroup *bg = new QButtonGroup(1, Horizontal, "Actions", this) ;
  //QGroupBox *bg = new QGroupBox(1, Horizontal, "Actions", this) ;
  QGroupBox *bg = new QGroupBox("Actions", this) ;
  bg->setAlignment( Qt::Alignment(Qt::AlignHCenter) );
  //QPushButton *clearall = new QPushButton( "Clear All", bg);
  QPushButton *clearall = new QPushButton( "Clear All", bg);
  connect( clearall, SIGNAL(clicked()),
	   SLOT(ClearAll()) );

  //QPushButton *selectall = new QPushButton( "Select All", bg);
  QPushButton *selectall = new QPushButton( "Select All");
  connect( selectall, SIGNAL(clicked()),
	   SLOT(SelectAll()) );

  //QPushButton *loadfile = new QPushButton( "Load File...", bg);
  QPushButton *loadfile = new QPushButton( "Load File...");
  connect( loadfile, SIGNAL(clicked()),
	   SLOT(ReadFile()) );


  //QPushButton *about = new QPushButton( "About...", bg);
  QPushButton *about = new QPushButton( "About...");
  connect( about, SIGNAL(clicked()),
	   SLOT(About()) );

  // add the QPushButtons to the QButtonGroup
  QVBoxLayout *vbuttons = new QVBoxLayout;
  //vbuttons->setSpacing(0);
  vbuttons->addWidget( clearall  );
  vbuttons->addWidget( selectall );
  vbuttons->addWidget( loadfile  );
  vbuttons->addWidget( about     );
  bg->setLayout( vbuttons );


  grid->addWidget(bg, 4, 1);
  // guessing at coordinates
  /*grid->addWidget( clearall , 4, 1);
  grid->addWidget( selectall, 4, 2);
  grid->addWidget( loadfile , 4, 3);
  grid->addWidget( about    , 4, 4);*/


  //m_timer = new QTimer(this, "mycwtimer");
  m_timer = new QTimer(this);
  
  connect(m_timer, SIGNAL(timeout()),
	   SLOT(playNextChar()) );

  //setCaption("CW Trainer");
  setWindowTitle("CW Trainer");
  setLayout( grid );
}

//this function is ugly...
//too many special cases and flags
void MyWidget::playNextChar()
{
  //file mode or randchar mode?
  bool dospace = false;
  bool play_char = true;
  char next_char;

  char prosign[2];
  prosign[0] = prosign[1] = 0;
  
  if(m_charpipe.isEmpty())
    {
      int j=(int) ((float)m_numInTest*rand()/(RAND_MAX+1.0));
      if(m_numingroup > 6)
	dospace = true;
      else if(m_numingroup < 2)
	dospace = false;
      else
	dospace = (2 == (int) ((float)4*rand()/(RAND_MAX+1.0))) && m_lastchar != ' ';

      m_numingroup++;

      next_char = m_charsInTest[j];
    }
  else
    {
      QChar ch = m_charpipe[m_numingroup];
      m_numingroup++;

      if(ch == '\0')
	{
	  EndCW();
	  return;
	}

      //next_char = (char)ch.upper();
      next_char = ch.toUpper().toAscii();

      //couple of translations?
      if(next_char == '-' || next_char == ':' || next_char == '\\')
	next_char = '/'; 
      

      if(next_char == '^') //prosign
	{
	  ch = m_charpipe[m_numingroup];
	  m_numingroup++;
	  if(ch == '\0')
	    {
	      EndCW();
	      return;
	    }
	  //prosign[0] = (char)ch.upper();
	  prosign[0] = ch.toUpper().toAscii();

	  ch = m_charpipe[m_numingroup];
	  m_numingroup++;
	  if(ch == '\0')
	    {
	      EndCW();
	      return;
	    }
	  //prosign[1] = (char)ch.upper();
	  prosign[1] = ch.toUpper().toAscii();
	}

      if(next_char == ' ')
	dospace = true;

      if(CWTest::GetCW(next_char) == " ")
	{
	  play_char = false;
	}
    }

  if(dospace == false)
    {
      m_lastchar = next_char;
      //m_cwtext->setText(m_cwtext->text() + QString(QChar(next_char)));
      m_cwtext->setText(m_cwtext->toPlainText() + QString(QChar(next_char)));

      if(play_char)
	m_cw->PlayCWNote(CWTest::GetCW(next_char).c_str());
      else if(prosign[1])
	{
	  m_cw->PlayCWNote(CWTest::GetCW(prosign[0]).c_str());
	  //m_cwtext->setText(m_cwtext->text() + QString(QChar(prosign[0])));
	  m_cwtext->setText(m_cwtext->toPlainText() + QString(QChar(prosign[0])));
	  m_cw->PlayCWNote(CWTest::GetCW(prosign[1]).c_str());
	  //m_cwtext->setText(m_cwtext->text() + QString(QChar(prosign[1])));
	  m_cwtext->setText(m_cwtext->toPlainText() + QString(QChar(prosign[1])));
	}

      //m_timer->changeInterval(m_charspace);
      m_timer->setInterval(m_charspace);
    }
  else
    {
      if(m_charpipe.isEmpty())
	m_numingroup = 0;

      m_lastchar = ' ';
      m_cwtext->setText(m_cwtext->toPlainText() + QString(" | "));
      //m_timer->changeInterval(m_wordspace);
      m_timer->setInterval(m_wordspace);
    }
}

void MyWidget::StartCW()
{
  m_timer->stop();
  //gather the checked boxes...
  memset(m_charsInTest, 0, 41);
  m_numInTest = 0;
  for(int i = 0 ; i < 40 ; i++)
    {
      QCheckBox *b = (QCheckBox*)m_lettergroup->find(i);
      if(b->isChecked())
	{
	  m_charsInTest[m_numInTest] = m_cwchars[i];
	  m_numInTest++;
	}
    }
  if(m_numInTest)
    {
      m_numingroup = 0;
      int speed = m_wpm->value();
      int space = m_spacing->value();

      NewSpeed(speed);
      NewSpacing(space);
      m_cwtext->setText(QString(""));
      m_timer->start(500);
    }
}

void MyWidget::NewSpeed(int speed)
{
  m_cw->AdjustSpeed(speed);
}

void MyWidget::NewSpacing(int space)
{
  m_charspace = (int)((((60000.0 / (float)space) / 7.0) / 3.0));
  m_wordspace = m_charspace * 2;
}

void MyWidget::EndCW()
{
  m_timer->stop();
  m_charpipe = "";
}

void MyWidget::About()
{
  QMessageBox::information( this
		 ,"About CW Trainer"
		 ,"CW Trainer 0.1.0"
		  "\nKoch and Farnsworth adaptable Morse Code trainer."
		  "\n\nBy John Wagner, N1QO and Jacob Anawalt, KD7YKO"
		  "\nhttp://sourceforge.net/projects/cwtrainer"
		  "\nGPL - Uses code from GWFON's Koch CW Trainer"
		  "\nSee README and LICENSE for more information");
}

void MyWidget::ClearAll()
{
  for(int i = 0 ; i < 40 ; i++)
    {
      QCheckBox *b = (QCheckBox*)m_lettergroup->find(i);
      if( b == NULL )
      {
         std::cerr << "Checkbox #" << i << " not found in MyWidget::SelectAll()\n";
         return;
      }
      b->setChecked(false);
    }
 
}

void MyWidget::SelectAll()
{
  for(int i = 0 ; i < 40 ; i++)
    {
      //QCheckBox *b = (QCheckBox*)m_lettergroup->find(i);
      int row, column;
      ConvertToPosition( i , &row, &column); // modifies row and column
      QCheckBox *b = (QCheckBox*)m_letterlayout->itemAtPosition(row, column);
      if( b == NULL )
      {
         std::cerr << "Checkbox #" << i << " not found in MyWidget::SelectAll()\n";
         return;
      }
      b->setChecked(true);
    }
 
}

void MyWidget::ReadFile()
{
  //QString s( QFileDialog::getOpenFileName( QString::null, "*", this ) );
  QString s( QFileDialog::getOpenFileName( this, "Open File", QString::null, "*" ) );
  if ( s.isEmpty() )
    return;

  QFile f(s);
  
  if ( f.open(QIODevice::ReadOnly) ) 
    {    // file opened successfully
      m_charpipe = "";
      QTextStream t( &f );        // use a text stream

      //while ( !t.eof() )         // until end of file...
      while ( !t.atEnd() )         // until end of file...
	{
	  QString s = t.readLine();       // line of text excluding '\n'
	  s += " ";
	  m_charpipe += s;
	  //printf( "%3d: %s\n", n++, (const char *)s );
	}
      f.close();
      StartCW();
    }
  else
    QMessageBox::information( this, "CW Trainer",
			      "An error occurred opening the file");

}

/**
* Takes a single integer and converts it to acoordinate on the grid of characters.
* The coordinate is passed back using the two reference paramaters, row and column.
*/
void MyWidget::ConvertToPosition( int i, int *row, int *column)
{
   const char *ptr = m_cwchars;
   int chars_in_row = QString(ptr).length() / CHARACTER_GRID_ROWS;
   *row = 0;
   if( i != 0) // take care of divide-by-zero
      *row = i / chars_in_row;
   *column  = i - (*row) * chars_in_row;
}

/**
* Takes a row and a column and converts it to a single index in the grid of characters.
*/
int MyWidget::ConvertFromPosition( int row, int column )
{
   const char *ptr = m_cwchars;
   int chars_in_row = QString(ptr).length() / CHARACTER_GRID_ROWS;
   return column + row * chars_in_row;
}
